### PR TITLE
Always move methods in flattener

### PR DESCRIPTION
**Pull request by aisamanra**
### Motivation
I came across an edge-case when deleting the old flattening pass: specifically, creating the `static-init` could un-flatten things in the situation that you had a method definition inside a send at the top-level.

For example, the new flattener would not have modified this file, because it did not contain any nested method definitions:
```ruby
foo def bar; end
def baz; end
```
However, when we create a `static-init` pass, we move everything that's not a definition into a method called `static-init`, which resulted in a file that looked like this:
```ruby
def <static-init>
  foo def bar; end
end
def baz; end
```
which meant we now have nested methods again! This PR modifies the flatten pass in a few small ways: firstly, it _always_ moves methods: the above file now gets flattened to
```ruby
foo :bar 
def bar; end
def baz; end
```
it also has logic to make sure that the visibility modifiers that we process and remove from the tree (`private`, `protected`, `public`) all move with the methods, as well.

Finally, because so many methods are moved now, I was finding that a _ton_ of exp files were changing because they were getting replaced with symbols or `EmptyTree`s. This changes it so that we don't have a ridiculous number of `EmptyTree`s from moving the methods.

### Test plan

See included automated tests.

### CODE REVIEW

File: rewriter/flatten.cc [hunk 1]
```diff
@@ -161,6 +161,10 @@ class FlattenWalk {
             }
             moveQueue[idx] = {move(expr), staticLevel};
         }
+
+        bool atTopLevel() const {
+            return stack.empty();
+        }
```
Comment (aisamanra) [inline rewriter/flatten.cc]: We  use this to determine whether to replace a moved `MethodDef` with a symbol or an `EmptyTree`. (This isn't strictly necessary, but it makes the exp files a lot cleaner.)

File: rewriter/flatten.cc [hunk 2]
```diff
@@ -313,14 +325,30 @@ class FlattenWalk {
...
+        if ((send.fun == core::Names::private_() ||
+             send.fun == core::Names::privateClassMethod() ||
+             send.fun == core::Names::protected_() ||
+             send.fun == core::Names::public_()) &&
+            send.args.size() == 1 &&
+            send.recv->isSelfReference()) {
+            return true;
+        }
+        return false;
+    }
```
Comment (aisamanra) [inline rewriter/flatten.cc]: These are the sends that we move: sigs and visibility modifiers. All the modifiers listed here we treat specially in the namer.

File: rewriter/flatten.cc [hunk 3]
```diff
@@ -351,8 +378,15 @@ class FlattenWalk {
...
-            // we'll replace MethodDefs with the symbol that corresponds to the name of the method
-            auto replacement = ast::MK::Symbol(methodDef->loc, methodDef->name);
+            unique_ptr<ast::Expression> replacement;
+
+            if (methods.atTopLevel()) {
+                replacement = ast::MK::EmptyTree();
+            } else {
+                // we'll replace MethodDefs with the symbol that corresponds to the name of the method
+                replacement = ast::MK::Symbol(methodDef->loc, methodDef->name);
+            }
```
Comment (aisamanra) [inline rewriter/flatten.cc]: This will only rewrite a method to a symbol if it's in a position where a symbol might be actually used (e.g. the `MethodDef` is an argument to a send.)

File: rewriter/flatten.cc (hunk 1 above)
Comment (DarkDimius) [reply rewriter/flatten.cc]: Let's make a comment that this is _syntactical_ top level rather than semantical and add a tests for 
```
puts(def foo; end;)
```
on top level

File: rewriter/flatten.cc [hunk 4]
```diff
@@ -313,14 +326,28 @@ class FlattenWalk {
         }
         classDef->rhs = addClassDefMethods(ctx, std::move(classDef->rhs), classDef->loc);
         return classDef;
-    };
+    }
+
+    bool shouldMoveSend(ast::Send &send) {
```
Comment (DarkDimius) [inline rewriter/flatten.cc]:
```suggestion
    bool shouldMoveSend(unique_ptr<ast::Send> &send) {
```

Comment (DarkDimius) [reply rewriter/flatten.cc]: Less explicit pointer dereferences & more uniform code.

File: rewriter/flatten.cc (hunk 1 above)
Comment (DarkDimius) [reply rewriter/flatten.cc]: (also, ~every comment left in PRs to explain something is worth leaving in source)

File: rewriter/flatten.cc (hunk 3 above)
Comment (DarkDimius) [reply rewriter/flatten.cc]: Seems like ad-hoc optimization to me, but sure.
CFG finalization passes before typechecking have this optimization as a proper pass.

**Comment by DarkDimius** [comment]
WDYT about removing the current ad-hoc optimization?

I agree with a sentiment of having simpler trees, but I'm concerned that this is done by an unrelated phase and thus not systematically(see empty trees left in exp files).
I think it's better done by a separate simplification/regularization pass that could be the last one in the set of Rewriter phases

**Comment by gdritter-stripe** [comment]
I think removing this ad-hoc optimization makes sense. I'll pare it back down to remove the top-level differences and the emptytree pass and then write a separate optimization pass afterwards.

File: rewriter/flatten.cc [hunk 5]
```diff
@@ -313,23 +323,40 @@ class FlattenWalk {
...
+        methods.sendDepth++;
         // we might want to move sigs, so we mostly use the same logic that we use for methods. The one exception is
         // that we don't know the 'staticness level' of a sig, as it depends on the method that follows it (whether that
         // method has a `self.` or not), so we'll fill that information in later
-        if (send->fun == core::Names::sig()) {
-            curMethodSet().pushScope(computeScopeInfo(ScopeType::StaticMethodScope));
+        if (shouldMoveSend(send)) {
+            auto scopeType =
+                send->fun == core::Names::sig() ? ScopeType::StaticMethodScope : ScopeType::VisibilityModScope;
+            methods.pushScope(computeScopeInfo(scopeType));
```
Comment (elliottt) [inline rewriter/flatten.cc]: Is it important that `computeScopeInfo` doesn't handle `VisibilityModScope` explicitly?

File: test/testdata/resolver/sig_misc.rb [hunk 6]
```diff
@@ -30,7 +30,7 @@ def f1(x) # error: Type not specified for argument
     T1.new
   end
 
-  sig do params(x: T1).returns(T2) end.hithere # error: Can not call method `hithere` on void type
+  sig do params(x: T1).returns(T2) end.hithere # error: Method `hithere` does not exist on `NilClass`
```
Comment (elliottt) [inline test/testdata/resolver/sig_misc.rb]: It's unfortunate that this error message changes, as `sig` doesn't return a `NilClass` result. This isn't blocking, but it would be great if we could get back to the old error.

File: rewriter/flatten.cc [hunk 7]
```diff
@@ -367,8 +371,15 @@ class FlattenWalk {
         auto &methods = curMethodSet();
         // if we get a MethodData back, then we need to move this and replace it
         if (auto md = methods.popScope()) {
-            // we'll replace MethodDefs with the symbol that corresponds to the name of the method
-            auto replacement = ast::MK::Symbol(methodDef->loc, methodDef->name);
+            unique_ptr<ast::Expression> replacement;
```
Comment (dmitry-stripe) [inline rewriter/flatten.cc]: let's extract this to a separate patch?

File: rewriter/flatten.cc [hunk 8]
```diff
@@ -313,14 +313,30 @@ class FlattenWalk {
         }
         classDef->rhs = addClassDefMethods(ctx, std::move(classDef->rhs), classDef->loc);
         return classDef;
-    };
+    }
+
+    bool shouldMoveSend(ast::Send& send) {
```
Comment (dmitry-stripe) [inline rewriter/flatten.cc]: that seems super bespoke to me.

File: rewriter/flatten.cc [hunk 9]
```diff
@@ -313,23 +323,40 @@ class FlattenWalk {
         }
         classDef->rhs = addClassDefMethods(ctx, std::move(classDef->rhs), classDef->loc);
         return classDef;
-    };
+    }
+
+    bool shouldMoveSend(unique_ptr<ast::Send> &send) {
```
Comment (dmitry-stripe) [inline rewriter/flatten.cc]: This seems super bespoke to me. Do I understand it right that meands we _don't_ actually  introduce an invariant?

**Comment by dmitry-stripe** [comment]
Oh second thought, what's the motivation behind this patch?
What would it enable?

>  creating the static-init could un-flatten things in the situation that you had a method definition inside a send at the top-level.

which phase would be doing that? 
If it's after namer/resolver: we don't care as it's past the critical point in the pipeline.
I don't expect _any noticeable wins from having CFGTyperAndCollector use a shallow tree mapper.

File: test/testdata/resolver/sig_misc.rb (hunk 6 above)
Comment (dmitry-stripe) [reply test/testdata/resolver/sig_misc.rb]: +1, could you elaborate on this more? Where does `NilClass` come from?

File: rewriter/flatten.cc [hunk 10]
```diff
@@ -352,7 +379,12 @@ class FlattenWalk {
         // if we get a MethodData back, then we need to move this and replace it
         if (auto md = methods.popScope()) {
             // we'll replace MethodDefs with the symbol that corresponds to the name of the method
-            auto replacement = ast::MK::Symbol(methodDef->loc, methodDef->name);
+            unique_ptr<ast::Expression> replacement;
+            if (methods.atClassTopLevel()) {
+                replacement = ast::MK::EmptyTree();
```
Comment (dmitry-stripe) [inline rewriter/flatten.cc]: Let's extract this to a separate PR.
I have FUD about this. Let's either be systematic and always use symbol or always use empty tree

File: test/testdata/rewriter/dsl_builder.rb.rewrite-tree.exp [hunk 11]
```diff
@@ -1,5 +1,7 @@
 class <emptyTree><<C <root>>> < ()
   class <emptyTree>::<C TestDSLBuilder><<C <todo sym>>> < (::<todo sym>)
+    <self>.dsl_optional(:"no_getter_or_setter", <emptyTree>::<C String>, {:"skip_getter" => true, :"skip_setter" => true})
```
Comment (dmitry-stripe) [inline test/testdata/rewriter/dsl_builder.rb.rewrite-tree.exp]: why did the order change?

File: rewriter/flatten.cc (hunk 9 above)
Comment (aisamanra) [reply rewriter/flatten.cc]: `shouldMoveSend` only applies to the set of things that we remove in the namer
(specifically [in this `postTransformSend`](https://github.com/sorbet/sorbet/blob/master/namer/namer.cc#L454)). If it did not, then we would just need to manage symbol visibility in a slightly different way.

File: test/testdata/resolver/sig_misc.rb (hunk 6 above)
Comment (aisamanra) [reply test/testdata/resolver/sig_misc.rb]: This sig gets moved along with the method underneath it and then is replaced with an `EmptyTree`, which evaluates to `NilClass`.

File: rewriter/flatten.cc (hunk 9 above)
Comment (DarkDimius) [reply rewriter/flatten.cc]: This seems to be a big crosstalk & action at a distance.
Let's either remove it(i'd prefer that) or document it & make both namer & resolver use the same helper method to detect it.

File: test/testdata/rewriter/dsl_builder.rb.rewrite-tree.exp (hunk 11 above)
Comment (aisamanra) [reply test/testdata/rewriter/dsl_builder.rb.rewrite-tree.exp]: This PR makes it so that method definitions and their attached sigs are always moved to the end of the class. This was not a method definition, so it did not get moved: hence why it is now here.

**Comment by ptarjan** [comment]
Hey! What's the state of this PR? If it is under active development, please let us know otherwise we should close it and re-open when we pick it back up.

**Comment by aisamanra** [comment]
I need a different approach for this and have not yet revisited it, but I'm going to close this PR as it's been contentious and big, and redo it fresh elsewhere.
